feat(file-mode-api: add filename extractor component

aldogonzalez8 opened the pull request:
Resolves https://github.com/airbytehq/airbyte-internal-issues/issues/12196

maxi297 commented on airbyte_cdk/sources/declarative/retrievers/file_uploader.py:
```diff
@@ -1,50 +1,68 @@
+            relative_path = relative_path.lstrip("/")
+            file_relative_path = Path(relative_path)
 
             full_path = files_directory / file_relative_path
```
Should we have the stream name somewhere in there? It feels like multiple streams could have a file with the same name.

Even more than this, should we have a unique ID per file? It feels like there could even be two files in the same stream with the same name...

maxi297 replied:
Ok so what we are saying is that it is the developer's responsibility to make sure there are no clash. Could we remove this concern from the developer's and do it ourselves? 

Regarding timing: I'm not 100% sure we need this right now and maybe we can make `filename_extractor` optional in the future when we find a solution this this. On the top of my head, I can only see one way and it is when the stream declares a PK which seems to be common when I checked for Confluence, Jira and Salesforce so maybe this is viable in the future

aldogonzalez8 replied:
> so what we are saying is that it is the developer's responsibility to make sure there are no clash. Could we remove this concern from the developer's and do it ourselves?

No, I didn't make myself clear. I'm sorry about that. To reduce the risk of collisions, I will add the stream name + unique ID on the backend (CDK).

maxi297 replied:
What is the logic for the unique ID? Autogenerated UUID?

aldogonzalez8 replied:
I think we can:

- Add ourselves stream name to the path reducing collision risk
- Make filename_extractor optional so the developer can include a unique ID. There is a risk that he could mess up, but we can add some documentation to the component. 
- Use Autogenerated UUID if filename_extractor is not present.

maxi297 reviewed (approved):
Accepting under the premise that we are fine that the connector developer ensure no file collisions for now

aldogonzalez8 commented:
> Accepting under the premise that we are fine that the connector developer ensure no file collisions for now

@maxi297  This is still true with slight modifications that we can refine in the future:

- We added the stream name to the path, reducing collision risk.
- Make filename_extractor optional so the developer can include a unique ID. There is a risk that dev could mess up the interpolation, but we added documentation to the component.
- Use Autogenerated UUID if filename_extractor is not present.
